### PR TITLE
[mdns] Split response packets if necessary.

### PR DESCRIPTION
**Pull request by romanb**
Currently, when an MDNS response gets larger than 9000 bytes due to the number of addresses being encoded, a panic is triggered. See https://github.com/libp2p/rust-libp2p/issues/1872. There are primarily two approaches for dealing with the `9000` bytes MDNS packet size limit: Creating multiple packets or trimming the packet by dropping addresses. Since I didn't find any restriction in the MDNS specifications w.r.t. the number of multicast response packets that can be sent (we don't support or send unicast response packets, I think), I opted for the former. So we now prevent MDNS response packets becoming too large (> 9000 bytes) by using conservative upper bounds and creating multi-packet responses if necessary (effectively when there are more than 29 addresses to encode, which effectively means when the local peer has more than 29 listening addresses, as a peer only responds with its own addresses). I'm only just starting out with studying the MDNS specifications, so feel free to set me straight if I'm doing anything nonsensical here.

Additionally, I opted for just skipping addresses that don't fit into a TXT record or contain invalid characters, instead of failing to send a response (which may contain other valid records), thereby logging a warning for the skipped addresses.

As such, closes https://github.com/libp2p/rust-libp2p/issues/1872.

### CODE REVIEW

File: protocols/mdns/src/dns.rs [hunk 1]
```diff
@@ -80,68 +101,70 @@ pub fn build_query() -> Vec<u8> {
...
-    append_u16(&mut out, addresses.len() as u16);
+    let peer_id_bytes = encode_peer_id(&peer_id);
+    debug_assert!(peer_id_bytes.len() <= 0xffff);
 
-    // Our single answer.
-    // The name.
-    append_qname(&mut out, SERVICE_NAME);
+    let mut packets = Vec::new();
+
+    'next_packet: loop {
```
Comment (mxinden) [inline protocols/mdns/src/dns.rs:124]: It took me a bit to understand this loop. The following control flow would be easier for me to understand. What do you think?

```diff
diff --git a/protocols/mdns/src/dns.rs b/protocols/mdns/src/dns.rs
index efb5a11a..fcc62fc2 100644
--- a/protocols/mdns/src/dns.rs
+++ b/protocols/mdns/src/dns.rs
@@ -120,38 +120,23 @@ pub fn build_query_response(
     debug_assert!(peer_id_bytes.len() <= 0xffff);
 
     let mut packets = Vec::new();
-
-    'next_packet: loop {
-        let mut records = Vec::with_capacity(addresses.len() * MAX_TXT_RECORD_SIZE);
-
-        // Try to pack as many records into an MDNS response packet
-        // as are known to fit comfortably.
-        for _ in 0 .. MAX_RECORDS_PER_PACKET {
-            if let Some(addr) = addresses.next() {
-                let txt_to_send = format!("dnsaddr={}/p2p/{}", addr.to_string(), peer_id.to_base58());
-                let mut txt_record = Vec::with_capacity(txt_to_send.len());
-                match append_txt_record(&mut txt_record, &peer_id_bytes, ttl, &txt_to_send) {
-                    Ok(()) => {
-                        records.push(txt_record);
-                        continue
-                    }
-                    Err(e) => {
-                        log::warn!("Excluding address {} from response: {:?}", addr, e);
-                    }
-                }
+    let mut records = Vec::with_capacity(addresses.len() * MAX_TXT_RECORD_SIZE);
+
+    while let Some(addr) = addresses.next() {
+        let txt_to_send = format!("dnsaddr={}/p2p/{}", addr.to_string(), peer_id.to_base58());
+        let mut txt_record = Vec::with_capacity(txt_to_send.len());
+        match append_txt_record(&mut txt_record, &peer_id_bytes, ttl, &txt_to_send) {
+            Ok(()) => {
+                records.push(txt_record);
+                continue
             }
-
-            // Push the final packet, if there are still unpacked records.
-            if !records.is_empty() {
-                packets.push(query_response_packet(id, &peer_id_bytes, &records, ttl));
+            Err(e) => {
+                log::warn!("Excluding address {} from response: {:?}", addr, e);
             }
-
-            break 'next_packet
         }
-
-        // Reached MAX_RECORDS_PER_PACKET. Finish up the current packet and
-        // begin a new one.
-        packets.push(query_response_packet(id, &peer_id_bytes, &records, ttl));
+        if records.len() == MAX_RECORDS_PER_PACKET || addresses.len() == 0 {
+            packets.push(query_response_packet(id, &peer_id_bytes, &records, ttl));
+        }
     }
```

File: protocols/mdns/src/dns.rs [hunk 2]
```diff
@@ -349,11 +394,8 @@ impl fmt::Display for MdnsResponseError {
             MdnsResponseError::TxtRecordTooLong => {
                 write!(f, "TXT record invalid because it is too long")
             }
-            MdnsResponseError::NonAsciiMultiaddr => write!(
-                f,
-                "A multiaddr contains non-ASCII characters when serializd"
-            ),
-            MdnsResponseError::ResponseTooLong => write!(f, "DNS response is too long"),
+            MdnsResponseError::NonAsciiMultiaddr =>
+                write!( f, "A multiaddr contains non-ASCII characters when serialized"),
```
Comment (mxinden) [inline protocols/mdns/src/dns.rs:398]:
```suggestion
                write!(f, "A multiaddr contains non-ASCII characters when serialized"),
```

**Review by mxinden — APPROVED** [review]
Thanks for the bug-fix as well as the overall improvements! From what I can tell, splitting responses into multiple packets is fine.

File: protocols/mdns/src/dns.rs (hunk 1 above)
Comment (romanb) [reply protocols/mdns/src/dns.rs]: What do you think about https://github.com/libp2p/rust-libp2p/pull/1877/commits/397b25026db0ea6103790c2b0873c4b36c0d5d40? With your suggestion as-is, I think the `continue` in the `Ok` branch would effectively render `MAX_RECORDS_PER_PEER` ineffective.

Comment (mxinden) [reply protocols/mdns/src/dns.rs]: Thanks, looks good to me. Indeed I missed the `continue`.

**Review by mxinden — APPROVED** [review]

**Pull request merged**